File: src/storeFx/play.ts
import * as Fx from '../fx/fx';
import type { Store } from '../store/types';
import { audio, replayGainValue, value } from './playerContext';

export function make() {
  let fx;

  if (value.contents == null) {
    const audioContext = new AudioContext();
    const replayGainNode = audioContext.createGain();
    const volumeNode = audioContext.createGain();

    audioContext
      .createMediaElementSource(audio)
      .connect(replayGainNode)
      .connect(volumeNode)
      .connect(audioContext.destination);

    fx = Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
      const state = store.getState();

      volumeNode.gain.value = state.player.muted ? 0 : state.player.volume;

      if (replayGainValue.contents) {
        replayGainNode.gain.setValueAtTime(
          replayGainValue.contents,
          audioContext.currentTime,
        );
      }

      value.contents = { audioContext, replayGainNode, volumeNode };

      return Fx.Ok();
    });
  } else {
    fx = Fx.Ok();
  }

  return Fx.bind(fx, () =>
    Fx.Async(() =>
      audio.play().then(
        () => Fx.Ok(),
        () => Fx.Err(undefined),
      ),
    ),
  );
}

File: src/storeFx/play.ml
open DomExtra
open Fx
open Monad_syntax

let make () =
  let* () =
    match !PlayerContext.value with
    | Some _ -> Ok ()
    | None ->
        let open WebAudio in
        let audioContext = AudioContext.make () in
        let replayGainNode = audioContext |. AudioContext.createGain in
        let volumeNode = audioContext |. AudioContext.createGain in
        audioContext
        |. AudioContext.createMediaElementSource PlayerContext.audio
        |. AudioNode.connect replayGainNode
        |. AudioNode.connect volumeNode
        |. AudioNode.connect (audioContext |. AudioContext.destination)
        |> ignore;
        let* Deps.{ store } = ask () in
        let state = store |. Zustand.getState in
        volumeNode |. AudioNode.gain |. AudioParam.setValue state.player.volume;
        !PlayerContext.replayGainValue
        |> Option.iter (fun replayGainValue ->
            replayGainNode |. AudioNode.gain
            |. AudioParam.setValueAtTime replayGainValue
                 (audioContext |. AudioContext.currentTime));
        PlayerContext.value := Some { audioContext; replayGainNode; volumeNode };
        Ok ()
  in
  Async
    (fun _ ->
      PlayerContext.audio |. HtmlAudioElement.play
      |> Js.Promise.then_ (fun () -> Ok () |> Js.Promise.resolve)
      |> Js.Promise.catch (fun _ -> Error () |> Js.Promise.resolve))
